man/man1/shread2.1
.\" Automatically generated by Pandoc 2.17.1.1
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "shread2" "1" "2021-02-15" "Fortran 95" "SHTOOLS 4.10"
.hy
.SH SHRead2
.PP
Read spherical harmonic coefficients from a CHAMP or GRACE-like
ascii-formatted file.
.SH Usage
.PP
call SHRead2 (\f[V]filename\f[R], \f[V]cilm\f[R], \f[V]lmax\f[R],
\f[V]gm\f[R], \f[V]r0_pot\f[R], \f[V]error\f[R], \f[V]dot\f[R],
\f[V]doystart\f[R], \f[V]doyend\f[R], \f[V]epoch\f[R],
\f[V]exitstatus\f[R])
.SH Parameters
.TP
\f[V]filename\f[R] : input, character(*)
The ascii-formatted filename containing the spherical harmonic
coefficients.
.TP
\f[V]cilm\f[R] : output, real(dp), dimension (2, \f[V]lmax\f[R]+1, \f[V]lmax\f[R]+1)
The spherical harmonic coefficients contained in \f[V]filename\f[R].
.TP
\f[V]lmax\f[R] : output, integer(int32)
The maximum spherical harmonic degree of \f[V]cilm\f[R].
.TP
\f[V]gm\f[R] : output, real(dp)
The mass-gravitational constant.
.TP
\f[V]r0_pot\f[R] : output, real(dp)
The reference radius of the potential coefficients.
.TP
\f[V]error\f[R] : output, optional, real(dp), dimension (2, \f[V]lmax\f[R]+1, \f[V]lmax\f[R]+1)
The errors corresponding to the spherical harmonic coefficients
\f[V]cilm\f[R].
.TP
\f[V]dot\f[R] : output, optional, real(dp), dimension (2, \f[V]lmaxout\f[R]+1, \f[V]lmaxout\f[R]+1)
The time derivatives of the spherical harmonic coefficients.
The dimension of this array can be smaller than \f[V]lmax+1\f[R].
.TP
\f[V]doystart\f[R] : output, optional, real(dp)
The starting time of the solution.
.TP
\f[V]doyend\f[R] : output, optional, real(dp)
The ending time of the solution
.TP
\f[V]epoch\f[R] : output, optional, real(dp)
The epoch time for the time derivates.
.TP
\f[V]exitstatus\f[R] : output, optional, integer(int32)
If present, instead of executing a STOP when an error is encountered,
the variable exitstatus will be returned describing the error.
0 = No errors; 1 = Improper dimensions of input array; 2 = Improper
bounds for input variable; 3 = Error allocating memory; 4 = File IO
error.
.SH Description
.PP
\f[V]SHRead2\f[R] will read spherical harmonic coefficients from a CHAMP
or GRACE-like ascii-formatted file into an array \f[V]cilm\f[R].
The errors and time derivatives associated with the coefficients will be
read if the optional arrays \f[V]error\f[R] and \f[V]dot\f[R] are
specified, respectively.
The staring and ending date of the solution are specified by the
optional parameters \f[V]doystart\f[R] and \f[V]doyend\f[R], and the
epoch of the time derivates is specified by the optional parameter
\f[V]epoch\f[R].
The maximum spherical harmonic degree is read from the file, and the
file does not need to be ordered by angular degree or order.
.PP
This routine does not read and output all parameters in the file.
Records that are read (and at least partially output) include:
\f[V]EARTH\f[R], \f[V]GGM\f[R], \f[V]SHM\f[R], \f[V]GRCOF2\f[R],
\f[V]GRDOTA\f[R], \f[V]CALSDV\f[R], \f[V]gfc\f[R], \f[V]gfct\f[R], and
\f[V]dot\f[R].
Comments specified by the record \f[V]CMMNT\f[R] will be print out to
the screen, as will the record names that are not currently implemented.
.PP
Each line of the file starts with a character string describing what
follows.
.IP \[bu] 2
\f[V]EARTH\f[R] or \f[V]GGM\f[R]: \f[V]GM\f[R], \f[V]R0_POT\f[R]
.IP \[bu] 2
\f[V]SHM\f[R]: Maximum spherical harmonic degree of file.
.IP \[bu] 2
\f[V]GRCOF2\f[R], \f[V]CALSDV\f[R], or \f[V]gfc\f[R]: spherical harmonic
coefficients, formatted as (\f[V]l\f[R], \f[V]m\f[R], \f[V]clm\f[R],
\f[V]slm\f[R]) or (\f[V]l\f[R], \f[V]m\f[R], \f[V]clm\f[R],
\f[V]slm\f[R], \f[V]clm_error\f[R], \f[V]slm_error\f[R]).
.SH See also
.PP
shread, shreadjpl


src/SHRead2.f95
subroutine SHRead2(filename, cilm, lmax, gm, r0_pot, error, dot, doystart, &
                   doyend, epoch, exitstatus)
!------------------------------------------------------------------------------
!
!   This program will read the file format of spherical harmonic
!   coefficients used by the CHAMP/GRACE group into standard arrays for
!   of Cilm and the error. Not all data for all fields are returned.
!
!   Calling Parameters
!
!       IN
!           filename    The name of the file.
!
!       OUT
!           lmax        Maximum spherical harmonic degree.
!           cilm        An array of the spherical harmonic coefficients.
!           gm          GM
!           R0_pot      Reference radius of gravity field
!
!       OPTIONAL
!           error       An array containing the error of the cilm coefficients.
!           dot         An array containing the assumed derivatives in time.
!                       of the low degree spherical harmonic coefficients.
!           doy1        Beginning date of the gravity solution.
!           doy2        Ending date of the gravity solution.
!           epoch       Epochal date for the time derivatives.
!           exitstatus  If present, instead of executing a STOP when an error
!                       is encountered, the variable exitstatus will be
!                       returned describing the error.
!                       0 = No errors;
!                       1 = Improper dimensions of input array;
!                       2 = Improper bounds for input variable;
!                       3 = Error allocating memory;
!                       4 = File IO error.
!
!   Copyright (c) 2005-2019, SHTOOLS
!   All rights reserved.
!
!------------------------------------------------------------------------------
    use ftypes

    implicit none

    character(*), intent(in) :: filename
    integer(int32), intent(out) :: lmax
    real(dp), intent(out) :: cilm(:,:,:), gm, r0_pot
    real(dp), intent(out), optional :: error(:,:,:), dot(:,:,:), doystart, &
                                       doyend, epoch
    integer(int32), intent(out), optional :: exitstatus
    integer(int32) :: temp_max, mmax, maxl_cilm, maxl_error, maxl_dot
    real(dp) :: error_scale, doy1, doy2, epoch_temp
    character(114) :: comment
    integer(int32) :: l, m, stat
    character(6) :: c
    real(dp) :: clm, slm, clmsig, slmsig

    if (present(exitstatus)) exitstatus = 0

    cilm = 0.0_dp

    doy1 = 0.0_dp
    doy2 = 0.0_dp
    epoch_temp = 0.0_dp

    if (size(cilm(:,1,1)) < 2) then
        print*, "Error --- SHRead2"
        print*, "CILM must be dimensioned (2, *, *)."
        print*, "Input array is dimensioned ", size(cilm(:,1,1)), &
                size(cilm(1,:,1)), size(cilm(1,1,:))
        if (present(exitstatus)) then
            exitstatus = 1
            return
        else
            stop
        end if

    end if

    maxl_cilm = min(size(cilm(1,:,1)) - 1, size(cilm(1,1,:)) - 1)

    if (present(error)) then
        if (size(error(:,1,1)) < 2) then
            print*, "Error --- SHRead2"
            print*, "ERROR must be dimensioned (2, *, *)."
            print*, "Input array is dimensioned ", size(error(:,1,1)), &
                    size(error(1,:,1)), size(error(1,1,:))
            if (present(exitstatus)) then
                exitstatus = 1
                return
            else
                stop
            end if

        end if

        error = 0.0_dp
        maxl_error = min(size(error(1,:,1)) - 1, size(error(1,1,:)) - 1)

    end if

    if (present(dot)) then
        if (size(dot(:,1,1)) < 2) then
            print*, "Error --- SHRead2"
            print*, "DOT must be dimensioned (2, *, *)."
            print*, "Input array is dimensioned ", size(dot(:,1,1)), &
                    size(dot(1,:,1)), size(dot(1,1,:))
            if (present(exitstatus)) then
                exitstatus = 1
                return
            else
                stop
            end if

        end if

        dot = 0.0_dp
        maxl_dot = min(size(dot(1,:,1)) - 1, size(dot(1,1,:)) - 1)

    end if

    open (13,file=filename)

    lmax = 0
    temp_max = 0
    gm = 0.0_dp
    r0_pot = 0.0_dp
    error_scale = 1.0_dp

    do 
        read(13,"(a6)", advance="no", iostat=stat) c
        if(stat /= 0) exit

        if (c == "EARTH " .or. c(1:3) == "GGM") then
            read(13,*) gm, r0_pot

        else if(c == "SHM   ") then 
            read(13,*) lmax, mmax, error_scale

            if (mmax /= lmax) then
                print*, "Warning --- SHREAD2"
                print*, "Maximum order is not equal to the maximum degree."
                print*, "LMAX = ", lmax
                print*, "MMax = ", mmax
            end if
            
            if (present(error) .and. error_scale /= 1.0_dp) then
                print*, "Warning --- SHREAD2"
                print*, "Standard deviations will be scale by ", error_scale
            end if

            if (size(cilm(1,:,1)) < lmax+1 .or. size(cilm(1,1,:)) < lmax+1) then
                print*, "Error ---SHRead2"
                print*, "CILM must be dimensioned (2, LMAX+1, LMAX+1) " // &
                        "where LMAX is ", lmax
                print*, "Input array is dimensioned ", size(cilm(1,:,1)), &
                        size(cilm(1,1,:)) 
                if (present(exitstatus)) then
                    exitstatus = 1
                    return
                else
                    stop
                end if

            else if (present(error)) then
                if (size(error(1,:,1)) < lmax+1 &
                        .or. size(error(1,1,:)) < lmax+1) then
                    print*, "Error ---SHRead2"
                    print*, "ERROR must be dimensioned (2, LMAX+1, " // &
                            "LMAX+1) where LMAX is ", lmax
                    print*, "Input array is dimensioned ", size(error(1,:,1)), &
                            size(error(1,1,:)) 
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if

            end if

        else if(c == "GRCOF2") then
            read(13,*) l, m, clm, slm, clmsig, slmsig, doy1, doy2

            if (l > maxl_cilm) then
                print*, "Error ---SHRead2"
                print*, "CILM must be dimensioned (2, LMAX+1, LMAX+1) " // &
                        "where LMAX is ", l
                print*, "Input array is dimensioned ", size(cilm(1,:,1)), &
                        size(cilm(1,1,:))
                if (present(exitstatus)) then
                    exitstatus = 1
                    return
                else
                    stop
                end if

            end if

            cilm(1,l+1,m+1) = clm
            cilm(2,l+1,m+1) = slm

            if (present(error)) then
                if(l > maxl_error) then
                    print*, "Error ---SHRead2"
                    print*, "ERROR must be dimensioned (2, " // &
                            "LMAX+1, LMAX+1) where LMAX is ", l
                    print*, "Input array is dimensioned ", size(error(1,:,1)), &
                            size(error(1,1,:)) 
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if
                
                if (error_scale /= 1.0_dp) then
                    error(1,l+1,m+1) = clmsig * error_scale
                    error(2,l+1,m+1) = slmsig * error_scale

                else
                    error(1,l+1,m+1) = clmsig
                    error(2,l+1,m+1) = slmsig

                end if

            end if

            if (l > temp_max) temp_max = l

        else if(c == "GRDOTA") then
            read(13,*) l, m, clm, slm, clmsig, slmsig, epoch_temp

            if (present(dot)) then
                if(l > maxl_dot) then
                    print*, "Error ---SHRead2"
                    print*, "DOT must be dimensioned (2, LMAX+1, LMAX+1) " // &
                            "where LMAX is ", l
                    print*, "Input array is dimensioned ", size(dot(1,:,1)), &
                            size(dot(1,1,:))
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if

                dot(1,l+1, m+1) = clm
                dot(2,l+1, m+1) = slm

            end if

            if (l > temp_max) temp_max = l

        else if(c == "CALSDV") then
            read(13,*) l, m, clm, slm

            if (l > maxl_cilm) then
                print*, "Error ---SHRead2"
                print*, "CILM must be dimensioned (2, LMAX+1, " // &
                        "LMAX+1) where LMAX is ", l
                print*, "Input array is dimensioned ", size(cilm(1,:,1)), &
                        size(cilm(1,1,:))
                if (present(exitstatus)) then
                    exitstatus = 1
                    return
                else
                    stop
                end if

            end if

            cilm(1,l+1,m+1) = clm
            cilm(2,l+1,m+1) = slm

            if (l > temp_max) temp_max = l

        else if(c == "gfc") then
            read(13,*) l, m, clm, slm, clmsig, slmsig

            if (l > maxl_cilm) then
                print*, "Error ---SHRead2"
                print*, "CILM must be dimensioned (2, LMAX+1, LMAX+1) " // &
                        "where LMAX is ", l
                print*, "Input array is dimensioned ", size(cilm(1,:,1)), &
                        size(cilm(1,1,:))
                if (present(exitstatus)) then
                    exitstatus = 1
                    return
                else
                    stop
                end if

            end if

            cilm(1,l+1,m+1) = clm
            cilm(2,l+1,m+1) = slm

            if (present(error)) then
                if(l > maxl_error) then
                    print*, "Error ---SHRead2"
                    print*, "ERROR must be dimensioned (2, LMAX+1, " // &
                            "LMAX+1) where LMAX is ", l
                    print*, "Input array is dimensioned ", size(error(1,:,1)), &
                            size(error(1,1,:))
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if

                if (error_scale /= 1.0_dp) then
                    error(1,l+1,m+1) = clmsig * error_scale
                    error(2,l+1,m+1) = slmsig * error_scale

                else
                    error(1,l+1,m+1) = clmsig
                    error(2,l+1,m+1) = slmsig

                end if

            end if

            if (l > temp_max) temp_max = l

        else if(c == "gfct") then
            read(13,*) l, m, clm, slm, clmsig, slmsig, epoch_temp

            if (l > maxl_cilm) then
                print*, "Error ---SHRead2"
                print*, "CILM must be dimensioned (2, LMAX+1, LMAX+1) " // &
                        "where LMAX is ", l
                print*, "Input array is dimensioned ", size(cilm(1,:,1)), &
                        size(cilm(1,1,:))
                if (present(exitstatus)) then
                    exitstatus = 1
                    return
                else
                    stop
                end if

            end if

            cilm(1,l+1,m+1) = clm
            cilm(2,l+1,m+1) = slm

            if (present(error)) then
                if (l > maxl_error) then
                    print*, "Error ---SHRead2"
                    print*, "ERROR must be dimensioned (2, LMAX+1, " // &
                            "LMAX+1) where LMAX is ", l
                    print*, "Input array is dimensioned ", size(error(1,:,1)), &
                            size(error(1,1,:)) 
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if

                if (error_scale /= 1.0_dp) then
                    error(1,l+1,m+1) = clmsig * error_scale
                    error(2,l+1,m+1) = slmsig * error_scale

                else
                    error(1,l+1,m+1) = clmsig
                    error(2,l+1,m+1) = slmsig

                end if

            end if

            if (l > temp_max) temp_max = l

        else if(c == "dot") then
            read(13,*) l, m, clm, slm, clmsig, slmsig

            if (present(dot)) then
                if(l > maxl_dot) then
                    print*, "Error ---SHRead2"
                    print*, "DOT must be dimensioned (2, LMAX+1, " // &
                            "LMAX+1) where LMAX is ", l
                    print*, "Input array is dimensioned ", size(dot(1,:,1)), &
                            size(dot(1,1,:))
                    if (present(exitstatus)) then
                        exitstatus = 1
                        return
                    else
                        stop
                    end if

                end if

                dot(1,l+1, m+1) = clm
                dot(2,l+1, m+1) = slm

            end if

            if (l > temp_max) temp_max = l

        else if(c == "CMMNT") then
            read(13,"(A114)") comment
            print*, "SHRead2 --- Ignoring comment : ", comment

        else
            read(13,*)
            print*, "SHRead2 --- Ignoring line ", c

        end if

    end do

    if (lmax == 0) lmax = temp_max

    if (present(doystart)) then
        doystart = doy1

    end if

    if (present(doyend)) then
        doyend = doy2

    end if

    if (present(epoch)) then
        epoch = epoch_temp

    end if

    close(13)

end subroutine SHRead2
